### PR TITLE
change UUID <-> Arrow mapping to (de)serialize to/from 16-byte FixedSizeBinary

### PR DESCRIPTION
**Pull request by jrevels**
...to make tables written out by Arrow.jl more portable. 

I'm dumb and didn't realize e.g. `pyarrow` doesn't support 128-bit ints, so currently UUID-containing tables will fail to be read by e.g. `pa.feather.read_feather(...)`.

There's even pyarrow documentation that [suggests to Python users that UUIDs can be written as FixedSizeBinary columns](https://arrow.apache.org/docs/python/extending_types.html), so this seems like a more friendly/correct way to do this.

### CODE REVIEW

**Comment by quinnj** [comment]
Failing tests?

**Comment by jrevels** [comment]
Ah, was missing some methods for FixedSizeListType types it seems. Fixed IIUC.

It's a shame that this entails a significant perf hit for `copy`:

```
julia> @time copy(uuids_uint128);
  0.002593 seconds (2 allocations: 3.248 MiB)

julia> length(uuids_fixed)
212886

julia> typeof(uuids_fixed)
Arrow.FixedSizeList{Base.UUID,Array{UInt8,1}}

julia> length(uuids_uint128)
212886

julia> typeof(uuids_uint128)
Arrow.Primitive{Base.UUID,Array{UInt128,1}}

julia> @time copy(uuids_fixed);
  0.167757 seconds (1.28 M allocations: 94.203 MiB, 2.81% gc time)

julia> @time copy(uuids_uint128);
  0.001690 seconds (2 allocations: 3.248 MiB)
```

I imagine we could optimize further to lessen/remove the overhead, though I think it'd preferable to do that in a follow-up

**Comment by jrevels** [comment]
Thanks to @SimonDanisch the UInt128 <-> NTuple{16,UInt8} cast employed here is now non-allocating 🔥

Reduces allocations for `copy(uuids_fixed)` a bit (see below) but looks like there's some additional performance still left on the table.

```
julia> @time copy(uuids_fixed);
  0.142450 seconds (1.06 M allocations: 74.713 MiB, 2.43% gc time)
```

I'll dig into it in a separate PR when I get a chance but AFAICT this is now good to go pending green CI.

**Comment by jrevels** [comment]
@quinnj This is now ready for review, recommend looking at #104 at the same time (ref https://github.com/JuliaData/Arrow.jl/pull/104#issuecomment-757369874, which I can take care of once either is merged) 🙂 thanks!

**Review by quinnj — APPROVED** [review]

**Comment by ericphanson** [comment]
BTW the PR title here says FixedSizeBinary, but what was implemented was FixedSizeList; why is that?

Asking since [arrow is considering a canonical extension for uuids that uses FixedSizeBinary](https://github.com/apache/arrow/pull/41299).

**Comment by jrevels** [comment]
> BTW the PR title here says FixedSizeBinary, but what was implemented was FixedSizeList; why is that?

IIRC back when this PR was authored, Arrow.jl would (de)serialize Julia values with type `FixedSizeList` w/ `UInt8` eltype as Arrow `FixedSizeBinary` values, and for w/e reason, that seemed like the right/easiest path to use in this PR (can't really remember the rationale for that, or whether there even was one 😅)

idk if that is still true nowadays though

here's the relevant `arrowtype` definition that existed when this PR was authored:

```
function arrowtype(b, x::FixedSizeList{T, A}) where {T, A}
    N = ArrowTypes.getsize(Base.nonmissingtype(T))
    if eltype(A) == UInt8
        Meta.fixedSizeBinaryStart(b)
        Meta.fixedSizeBinaryAddByteWidth(b, Int32(N))
        return Meta.FixedSizeBinary, Meta.fixedSizeBinaryEnd(b), nothing
    else
        children = [fieldoffset(b, "", x.data)]
        Meta.fixedSizeListStart(b)
        Meta.fixedSizeListAddListSize(b, Int32(N))
        return Meta.FixedSizeList, Meta.fixedSizeListEnd(b), children
    end
end
```

**Pull request merged**
